recipes/lora_finetune_distributed.py

```diff
@@ -451,22 +429,21 @@ def save_checkpoint(
         intermediate_checkpoint = epoch + 1 < self.total_epochs
         # To prevent GPU memory from spiking during checkpoint save,
         # we consolidate the full model and optim state dicts on CPU for rank 0
-        with FSDP.state_dict_type(
-            self._model,
-            StateDictType.FULL_STATE_DICT,
-            FullStateDictConfig(offload_to_cpu=True, rank0_only=True),
-            FullOptimStateDictConfig(offload_to_cpu=True, rank0_only=True),
-        ):
-            cpu_state_dict = self._model.state_dict()
-            if intermediate_checkpoint:
-                opt_state_dict = FSDP.optim_state_dict(self._model, self._optimizer)
-            else:
-                opt_state_dict = None
+
+        cpu_state_dict = utils.get_full_model_state_dict(
+            self._model, self._is_rank_zero
+        )
+
+        if intermediate_checkpoint:
+            opt_state_dict = utils.get_full_optimizer_state_dict(
+                self._optimizer, self._is_rank_zero
+            )
```

Comment:
@kartikayk `get_full_optimizer_state_dict` is converting sharded opotimizer state dict to full state dict on rank0. let me know if the UX looks smooth enough

Reply:
I think we should just use `torch.distributed.checkpoint.state_dict` instead of reinventing another API.

Reply:
talked with @fegin. will update PR soon to use `torch.distributed.checkpoint.state_dict` instead of `get_full_optimizer_state_dict`

Reply:
@fegin DCP to use `key=fqn` instead of `key=int` in `opt_state_dict['param_groups'][0].keys()`. would `key=fqn` fails `torch.optim.Optimizer.load_state_dict`? Trying to see if implementing rank0 load + boradcast in `torch.distributed.checkpoint.state_dict` become a hard blocker for this PR

Reply:
@fegin perf-wise, seems DCP is doing N gather if there are N params? `torch/distributed/_state_dict_utils.py`? want to double confirm since we were thinking of bucketing/coalescing to do fewer collectives on larger tensors (not an immeidate priority)

Reply:
@weifengpy For your first question, I think there should be a complementary API `set_state_dict` to call to modify the optimizer state dict before `torch.optim.Optimizer.load_state_dict`.